### PR TITLE
Add custom section with version info to WASM

### PR DESCRIPTION
**Pull request by mitjat**
The PR adds a custom section named `oasis_version` to each WASM. The contents are a JSON with keys "sha" (of the repo at the point when the service was built) and "serviceName" (intended for humans; derived from filename of wasm).

The goal is to make it easier to debug the many version mismatches we have seen recently.

This change is backwards compatible.

### CODE REVIEW

File: src/subcommands/build.rs [hunk 1]
```diff
@@ -217,6 +227,30 @@ pub fn prep_wasm(input_wasm: &Path, output_wasm: &Path, debug: bool) -> Result<(
         }
     }
 
+    // Add a section with version info for current git repo.
+    let git_sha = match Command::new("git").args(&["rev-parse", "HEAD"]).output() {
+        Err(_) => b"(git rev-parse failed)".to_vec(),
+        Ok(output) => strip_trailing_newline(output.stdout),
```
Comment (nhynes) [inline src/subcommands/build.rs:233]:
```suggestion
        Ok(output) => strip_trailing_newline(output.stdout),
        Err(_) => b"(git rev-parse failed)".to_vec(),
```

File: src/subcommands/build.rs [hunk 2]
```diff
@@ -217,6 +227,30 @@ pub fn prep_wasm(input_wasm: &Path, output_wasm: &Path, debug: bool) -> Result<(
...
+        .output()
+    {
+        Err(_) => false,
+        Ok(output) => !strip_trailing_newline(output.stdout).is_empty(),
+    };
+    module.customs.add(walrus::RawCustomSection {
+        name: "oasis_version".to_string(),
+        data: format!(
+            r#"{{"sha":"{}{}","serviceName":"{}"}}"#,
+            str::from_utf8(&git_sha[..])?,
```
Comment (nhynes) [inline src/subcommands/build.rs:246]:
```suggestion
            String::from_utf8(git_sha)?,
```

File: src/subcommands/build.rs [hunk 3]
```diff
@@ -217,6 +227,30 @@ pub fn prep_wasm(input_wasm: &Path, output_wasm: &Path, debug: bool) -> Result<(
...
+        Err(_) => b"(git rev-parse failed)".to_vec(),
+        Ok(output) => strip_trailing_newline(output.stdout),
+    };
+    let is_git_dirty = match Command::new("git")
+        .args(&["status", "--porcelain"])
+        .output()
+    {
+        Err(_) => false,
+        Ok(output) => !strip_trailing_newline(output.stdout).is_empty(),
+    };
```
Comment (nhynes) [inline src/subcommands/build.rs:241]:
```suggestion
    let is_git_dirty = match Command::new("git")
        .args(&["status", "--porcelain"])
        .output()
        .map(|o| !strip_trailing_newline(o.stdout).is_empty())
        .unwrap_or_default();
```

File: src/subcommands/build.rs [hunk 4]
```diff
@@ -217,6 +227,30 @@ pub fn prep_wasm(input_wasm: &Path, output_wasm: &Path, debug: bool) -> Result<(
         }
     }
 
+    // Add a section with version info for current git repo.
+    let git_sha = match Command::new("git").args(&["rev-parse", "HEAD"]).output() {
+        Err(_) => b"(git rev-parse failed)".to_vec(),
+        Ok(output) => strip_trailing_newline(output.stdout),
+    };
+    let is_git_dirty = match Command::new("git")
```
Comment (nhynes) [inline src/subcommands/build.rs:235]:
```suggestion
    let git_has_dirty_index = match Command::new("git")
```

File: src/subcommands/build.rs [hunk 5]
```diff
@@ -217,6 +227,30 @@ pub fn prep_wasm(input_wasm: &Path, output_wasm: &Path, debug: bool) -> Result<(
...
+    module.customs.add(walrus::RawCustomSection {
+        name: "oasis_version".to_string(),
+        data: format!(
+            r#"{{"sha":"{}{}","serviceName":"{}"}}"#,
+            str::from_utf8(&git_sha[..])?,
+            if is_git_dirty { " (DIRTY)" } else { "" },
+            input_wasm.file_stem().unwrap_or_default().to_string_lossy()
+        )
+        .as_bytes()
+        .to_vec(),
```
Comment (nhynes) [inline src/subcommands/build.rs:251]:
```suggestion
        .into_bytes(),
```

File: src/subcommands/build.rs [hunk 6]
```diff
@@ -189,6 +191,14 @@ fn build_rust_app(target: &Target, opts: &BuildOptions) -> Result<()> {
     Ok(())
 }
 
+/// Remove a trailing newline from a byte string.
+fn strip_trailing_newline(mut input: Vec<u8>) -> Vec<u8> {
```
Comment (nhynes) [inline src/subcommands/build.rs:195]: a helper function goes below the method that uses it

**Review by nhynes — APPROVED** [review]

File: src/subcommands/build.rs (hunk 6 above)
Comment (mitjat) [reply src/subcommands/build.rs:195]: Huh, ok. Funky convention; makes you zigzag when you read. :-1:

**Pull request merged**